pkg/subscribe/watcher_test.go

```diff
@@ -0,0 +1,214 @@
+	"fmt"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/rancher/apiserver/pkg/types"
+	"github.com/rancher/wrangler/pkg/schemas"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_stream(t *testing.T) {
+	tests := []struct {
+		name           string
+		sub            Subscribe
+		hasAccess      bool
+		wantStartEvent types.APIEvent
+		wantError      bool
+	}{
+		{
+			name: "stream all",
+			sub: Subscribe{
+				ResourceType: "watchable-resource",
+			},
+			hasAccess: true,
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "watchable-resource",
+			},
+		},
+		{
+			name: "stream by namespace",
+			sub: Subscribe{
+				ResourceType: "watchable-resource",
+				Namespace:    "test-ns",
+			},
+			hasAccess: true,
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "watchable-resource",
+				Namespace:    "test-ns",
+			},
+		},
+		{
+			name: "stream by selector",
+			sub: Subscribe{
+				ResourceType: "watchable-resource",
+				Selector:     "foo=bar",
+			},
+			hasAccess: true,
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "watchable-resource",
+				Selector:     "foo=bar",
+			},
+		},
+		{
+			name: "stream by id",
+			sub: Subscribe{
+				ResourceType: "watchable-resource",
+				ID:           "test-resource",
+			},
+			hasAccess: true,
+			wantStartEvent: types.APIEvent{
+				Name:         "resource.start",
+				ResourceType: "watchable-resource",
+				ID:           "test-resource",
+			},
+		},
+		{
+			name: "missing schema error",
+			sub: Subscribe{
+				ResourceType: "notaresource",
+			},
+			hasAccess: true,
+			wantError: true,
+		},
+		{
+			name: "unsupported schema error",
+			sub: Subscribe{
+				ResourceType: "listonly-resource",
+			},
+			hasAccess: true,
+			wantError: true,
+		},
+		{
+			name: "forbidden schema error",
+			sub: Subscribe{
+				ResourceType: "watchable-resource",
+			},
+			hasAccess: false,
+			wantError: true,
+		},
+	}
+	ws := WatchSession{
+		apiOp: &types.APIRequest{
+			Name: "test",
+			Schemas: &types.APISchemas{
+				Schemas: map[string]*types.APISchema{
+					"watchable-resource": {
+						Schema: &schemas.Schema{
+							ID: "watchable-resource",
+						},
+						Store: &mockStore{},
+					},
+					"listonly-resource": {
+						Schema: &schemas.Schema{
+							ID: "listonly-resource",
+						},
+					},
+				},
+			},
+			Request: &http.Request{},
+		},
+		getter: DefaultGetter,
+	}
+	for _, test := range tests {
+		ws.apiOp.AccessControl = &mockAC{hasAccess: test.hasAccess}
+		t.Run(test.name, func(t *testing.T) {
+			result := make(chan types.APIEvent, 1)
+			err := ws.stream(context.TODO(), test.sub, result)
+			if test.wantError {
+				assert.NotNil(t, err)
+				return
+			}
+			assert.Nil(t, err)
+			done := make(chan bool)
+			var gotEvent types.APIEvent
+			go func() {
+				gotEvent = <-result
+				done <- true
+			}()
+			select {
+			case <-done:
```

Comment:
```suggestion
			var gotEvent types.APIEvent
			select {
			case gotEvent = <-result:
```
I was able to try this out, and the test still runs since the separate go routine is not running any code. It is just blocking as well.

Separately this should not deadlock since you are using a buffered channel. 
If someone were to edit the implementation for `ws.stream(...)` to send multiple events on start, then we will hit a deadlock waiting for `ws.stream(...)` to return.

Reply:
You're right, the buffered channel made this unnecessary. I had needed it when I was trying to make it work with an unbuffered channel. Updated.